airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/write/object_storage/PartToObjectAccumulator.kt

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.load.write.object_storage
+
+import io.airbyte.cdk.load.command.DestinationStream
+import io.airbyte.cdk.load.file.object_storage.ObjectStorageClient
+import io.airbyte.cdk.load.file.object_storage.PartMetadataAssembler
+import io.airbyte.cdk.load.file.object_storage.RemoteObject
+import io.airbyte.cdk.load.file.object_storage.StreamingUpload
+import io.airbyte.cdk.load.message.Batch
+import io.airbyte.cdk.load.message.object_storage.IncompletePartialUpload
+import io.airbyte.cdk.load.message.object_storage.LoadablePart
+import io.airbyte.cdk.load.message.object_storage.LoadedObject
+import io.airbyte.cdk.load.state.object_storage.ObjectStorageDestinationState
+import io.airbyte.cdk.load.util.setOnce
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlinx.coroutines.CompletableDeferred
+
+class PartToObjectAccumulator<T : RemoteObject<*>>(
+    private val stream: DestinationStream,
+    private val client: ObjectStorageClient<T>,
+) {
+    private val log = KotlinLogging.logger {}
+
+    data class UploadInProgress<T : RemoteObject<*>>(
+        val streamingUpload: CompletableDeferred<StreamingUpload<T>> = CompletableDeferred(),
```

Comment:
Why does this need to be a deferred?

Reply:
Demoted to an atomic reference, now that I have the `hasStarted.setOnce()` guard.